### PR TITLE
[DBeX-60979] wrap 4142 3rd party action in logging

### PR DESCRIPTION
**Pull request by SamStuckey**
## Summary

- Update `Logging::ThirdPartyTransaction` logging module to allow for instance level method evaluation and result logging
- Update related logging service spec to include more robust examples with
    - silently failing method level logging (explicit and implicit error raising)
    - PORO (plain old ruby object) inclusion
- Wrap `Form526Submission#submit_form_4142` in logging
- Refactor and wrap central mail 3PI call from within the `CentralMail::SubmitForm4142Job` in logging

## Related Issues
- [ticket](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/60984)
- [epic](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/60952)

## Testing done

- There are no user facing test to do.  Only observable changes are in logging.
- Logging has been tested against existing and new specs
- extended Objects core functionality are covered with existing specs

## What areas of the site does it impact?
Form 526 Submission to EVSS (background worker, non-user facing)

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  [Documentation has been updated](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/implementation/reusable_logging_with_third_party_transaction_logger.md)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature (N/A)

### CODE REVIEW

File: app/controllers/v0/upload_supporting_evidences_controller.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ class UploadSupportingEvidencesController < ApplicationController
 
     FORM_ATTACHMENT_MODEL = SupportingEvidenceAttachment
 
-    wrap_with_logging :save_attachment_to_cloud!, additional_logs: {
+    wrap_with_logging :save_attachment_to_cloud!, additional_class_logs: {
```
Comment (SamStuckey) [inline app/controllers/v0/upload_supporting_evidences_controller.rb:12]: update to new syntax

File: app/workers/central_mail/submit_form4142_job.rb [hunk 2]
```diff
@@ -54,6 +67,14 @@ def perform(submission_id)
 
     private
 
+    def processor
+      @processor ||= EVSS::DisabilityCompensationForm::Form4142Processor.new(submission, jid)
+    end
+
+    def upload_to_central_mail
```
Comment (SamStuckey) [inline app/workers/central_mail/submit_form4142_job.rb:74]: refactor to make method wrappable

Comment (navaeddie) [reply app/workers/central_mail/submit_form4142_job.rb:74]: Not sure what your comment means here. Are you explaining you refactored upload_to_central_mail so some other method could wrap it?

Comment (SamStuckey) [reply app/workers/central_mail/submit_form4142_job.rb:74]: weird.. this comment doesn't belong here.  not sure where it came from

Comment (SamStuckey) [reply app/workers/central_mail/submit_form4142_job.rb:74]: ah sorry, I'm explaining why I created this method.  It was a refactor of the original logic to allow use of `wrap_with_logging`

File: app/workers/evss/document_upload.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ class EVSS::DocumentUpload
     :perform_initial_file_read,
     :perform_document_upload_to_evss,
     :clean_up!,
-    additional_logs: {
+    additional_class_logs: {
```
Comment (SamStuckey) [inline app/workers/evss/document_upload.rb:18]: update to new syntax

File: app/workers/central_mail/submit_form4142_job.rb [hunk 4]
```diff
@@ -4,9 +4,23 @@
...
 require 'evss/disability_compensation_form/metrics'
 require 'evss/disability_compensation_form/form4142_processor'
+require 'logging/third_party_transaction'
 
 module CentralMail
   class SubmitForm4142Job < EVSS::DisabilityCompensationForm::Job
+    extend Logging::ThirdPartyTransaction::MethodWrapper
+
+    # this is required to make instance variables available to logs via
+    # thw wrap_with_logging method
```
Comment (navaeddie) [inline app/workers/central_mail/submit_form4142_job.rb:14]: nit pick, I think "thw" is a typo and not some new fangled abbreviation all the cool coders are using these days?

File: lib/logging/third_party_transaction.rb [hunk 5]
```diff
@@ -61,14 +70,35 @@ def log_3pi_complete(method_name, additional_logs = {}, args = [])
...
+        # or
+        #
+        # SAD PATH (safe)
+        # { user_uuid: [:user_account, :id, :non_working_method, :something_else_that_breaks] }
+        # will be translated to
+        # { user_uuid: nil }
+        {}.tap do |obj|
+          additional_instance_logs.each do |key, method_chain|
+            # using :try ensures we can fail quietly
+            obj[key] = method_chain.inject(self, :try)
```
Comment (navaeddie) [inline lib/logging/third_party_transaction.rb:98]: What will happen if/when we fail quietly here?

**Review by navaeddie — APPROVED** [review]
This looks like a good clean refactor to me. I raised a question or two out of curiosity.

File: lib/logging/third_party_transaction.rb (hunk 5 above)
Comment (SamStuckey) [reply lib/logging/third_party_transaction.rb:98]: we will just log a nil value.  This is a feature, allowing people to pass method chains that may be unstable due to context, such as inheritance.  Also, just a safe gaurd so logging doesn't break important actions

File: app/workers/central_mail/submit_form4142_job.rb (hunk 4 above)
Comment (SamStuckey) [reply app/workers/central_mail/submit_form4142_job.rb]: fixed : )

File: lib/logging/third_party_transaction.rb (hunk 5 above)
Comment (SamStuckey) [reply lib/logging/third_party_transaction.rb:98]: Note that logging a nil value for one method chain will *not* affect the other log values

**Review by navaeddie — APPROVED** [review]
LGTM!

File: app/models/form526_submission.rb [hunk 6]
```diff
@@ -2,11 +2,25 @@
...
+  wrap_with_logging(
+    :submit_form_4142,
+    additional_class_logs: {
+      action: 'Begin as anciliary 526 submission'
+    },
+    additional_instance_logs: {
+      saved_claim_id: %i[saved_claim id],
+      user_uuid: %i[user_uuid]
+    }
+  )
```
Comment (SamStuckey) [inline app/models/form526_submission.rb:22]: Add logging per the ticket requirement

File: app/workers/central_mail/submit_form4142_job.rb [hunk 7]
```diff
@@ -4,9 +4,23 @@
...
+    # this is required to make instance variables available to logs via
+    # the wrap_with_logging method
+    attr_accessor :submission_id
+
+    wrap_with_logging(
+      :upload_to_central_mail,
+      additional_instance_logs: {
+        submission_id: [:submission_id]
+      }
+    )
```
Comment (SamStuckey) [inline app/workers/central_mail/submit_form4142_job.rb:22]: add logging per the ticket requirement

File: app/workers/central_mail/submit_form4142_job.rb [hunk 8]
```diff
@@ -34,12 +47,14 @@ class CentralMailResponseError < Common::Exceptions::BackendServiceException
     # @param submission_id [Integer] the {Form526Submission} id
     #
     def perform(submission_id)
+      @submission_id = submission_id
```
Comment (SamStuckey) [inline app/workers/central_mail/submit_form4142_job.rb:50]: makes this value available in the logging service

**Review by holdenhinkle — APPROVED** [review]
Nice!

**Pull request merged**